semeio/workflows/ahm_analysis/ahmanalysis.py

```diff
@@ -0,0 +1,456 @@
+    list_of_files = [
+        os.path.join(update_log_file, f) for f in os.listdir(update_log_file)
+    ]
+    if len(list_of_files) > 1:
+        raise OSError("ERROR more than one update_log_file.")
+    # read file
+    updatelog = pd.read_csv(
+        list_of_files[0],
+        delim_whitespace=True,
+        skiprows=6,
+        usecols=[0, 2, 3, 5, 6, 8, 10],
+        header=None,
+    )
+    # define header
+    updatelog.columns = [
+        "obs_number",
+        "obs_key",
+        "obs_mean",
+        "obs_std",
+        "status",
+        "sim_mean",
+        "sim_std",
+    ]
+    # drop last line ('=====...====')
+    # drop last row i.e. end of file line
+    updatelog.drop(updatelog.tail(1).index, inplace=True)
+    # ---------------------------------
+    # add proper name for obs_key in rows where value is '...'
+    replace = "..."
+    for i in range(len(updatelog.index)):
+        obskey_val = updatelog.at[i, "obs_key"]
+        if obskey_val == replace:
+            updatelog.at[i, "obs_key"] = key
+
+    return updatelog
+
+
+def make_obs_groups(key_map):
+    """Create a mapping of observation groups, the names will be:
+    data_key -> [obs_keys] and All_obs-{missing_obs} -> [obs_keys]
+    and All_obs -> [all_obs_keys]
+    """
+    combinations = key_map.copy()
+    if len(combinations) == 1:
+        return combinations
+    # We want all observations and all observations
+    # minus individual observations:
+    for combination_length in (len(key_map.keys()) - 1, len(key_map.keys())):
+        if combination_length == 1:
+            # We have hit a case with two observation groups, now we only
+            # need individual group and all obs.
+            continue
+        for subset in itertools.combinations(key_map.keys(), combination_length):
+            obs_group = list(
+                itertools.chain.from_iterable([key_map[key] for key in subset])
+            )
+            if combination_length == len(key_map.keys()) - 1:
+                missing_obs = [x for x in key_map.keys() if x not in set(subset)]
+                assert len(missing_obs) == 1
+                name = f"All_obs-{missing_obs[0]}"
+            else:
+                name = "All_obs"
+            combinations[name.replace(":", "_")] = obs_group
+    return combinations
+
+
+def list_active_observations(df_update_log):
+
+    """To get the inactive vs total observation info."""
+    df_active = df_update_log[(df_update_log["status"] == "Active")]
+    return str(len(df_active)) + " active/" + str(len(df_update_log.index))
+
+
+def create_path(runpathdf, target_name):
+    """to create path for output data"""
+    return tempfile.mkdtemp()
+
+
+def list_observations_misfit(obs_keys, df_update_log, misfit_df):
+    """To get the misfit for total observations (active/inactive)."""
+    misfit_values = []
+    total_obs_nr = len(df_update_log[df_update_log.status.isin(["Active", "Inactive"])])
```

Comment:
Can the status be anything other than `Active` or `Inactive`?

Reply:
yes  it can be Missing if the point/data is outside the grid/has no simulated data.